### PR TITLE
[LA.UM.7.1.r1] media: sde: rotator: Remove legacy swts code without hwts guard

### PR DESCRIPTION
**Pull request by MarijnS95**
The rotator driver attained hardware timestamp support on 4.14 but the
ported quirks for sdm6xx were not hidden behind this feature flag. This
meant a rogue sde_hw_rotator_swts_map would have executed on platforms
with hwts causing undefined behaviour such as panics due to accessing
uninitialized pointers, as sde_hw_rotator_swts_create is never called.

Merging the logic together causes sde_hw_rotator_swts_map to become
superfluous as sde_hw_rotator_swts_create - which maps the buffer too -
is always called when no HW timestamp support is available and the
swts_buf is not yet mapped.

Also remove the unused cpu-mapped swts_buffer from ported code,
which has been removed upstream from sde_hw_rotator_swts_create too [1].
Note that the struct and print statements retain this legacy field.

[1]: 6e2d3ce466b50ca43124f9e62177feefe9647bb5

---

This fixes a crash observed when opening Open Camera on Kumano, to test freshly added camera support in https://github.com/sonyxperiadev/kernel/pull/2196

### CODE REVIEW

**Comment by kholk** [comment]
Eh, are you sure you're not breaking Nile?

Actually, the issue is that this is a misport... needs to be fixed properly....

**Comment by MarijnS95** [comment]
> Eh, are you sure you're not breaking Nile?

I did not manage to break it (Discovery, Mermaid) with these changes. That is, I don't see how `sde_hw_rotator_free_ext` is triggered to share the SMMU with the MDP as indicated by the original port 6e2d3ce466b50ca43124f9e62177feefe9647bb5.

> Actually, the issue is that this is a misport... needs to be fixed properly....

I think I see what you mean. The code in 4.9 only creates swts resources when `iclient` is NULL, which is initialized once and never unset. Meaning a successive `sde_hw_rotator_free_ext` + `sde_hw_rotator_alloc_ext` only unmaps and remaps the dma_buf without ever detaching it in `sde_hw_rotator_swts_destroy`.

https://github.com/sonyxperiadev/kernel/blob/dbfc0e468e4edb9dec86c4dda75ddfc3b0b3f626/drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c#L2443-L2453

On 4.14 the same happens, except that there is no `iclient` anymore. Instead, `sde_hw_rotator_alloc_ext` only checks if the buffer is unmapped, and reattaches the entire thing which is an artifact from [other](https://github.com/sonyxperiadev/kernel/commit/6e2d3ce466b50ca43124f9e62177feefe9647bb5) [changes](https://github.com/sonyxperiadev/kernel/commit/0931849cf9c764bde12cf8cf8c1960c19297a82b).

https://github.com/sonyxperiadev/kernel/blob/d4dc6670ed82258d2f0c068feac15588f39b369d/drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c#L2762-L2774

This likely causes all sorts of breakage, so I think you want me to guard `sde_hw_rotator_swts_create` against attaching and pinning the smmu dma_buf again when it was not unpinned and detached first?

**Comment by kholk** [comment]
Please keep this PR open for another 2-3 weeks....

**Comment by ix5** [comment]
Merge conflict, is this PR still relevant?

**Comment by MarijnS95** [comment]
It does some vital cleanup but has been "super"seded by another patch. I'll see if I can find time to rebase those changes off, leaving only the cleanup.

**Comment by MarijnS95** [comment]
Force-pushed as requested. The first commit is relevant, the second one... Perhaps we should keep cleanup in `sde_hw_rotator_swts_{create,destroy}` but omit lazy creation in `sde_hw_rotator_swts_create` which was sort of superseded by #2234.

**Comment by pablomh** [comment]
Hi @MarijnS95. This has been in my list of patches to apply and hasn't introduced any noticeable regression in my pioneer (nile platform).

File: drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c [hunk 1]
```diff
@@ -2449,28 +2449,36 @@ static int sde_hw_rotator_swts_create(struct sde_hw_rotator *rot)
...
 
 	sde_smmu_ctrl(1);
 
-	data->srcp_attachment = sde_smmu_dma_buf_attach(data->srcp_dma_buf,
-			&rot->pdev->dev, SDE_IOMMU_DOMAIN_ROT_UNSECURE);
-	if (IS_ERR_OR_NULL(data->srcp_attachment)) {
-		SDEROT_ERR("sde_smmu_dma_buf_attach error\n");
-		rc = -ENOMEM;
-		goto err_put;
+	if (!data->srcp_attachment) {
```
Comment (kholk) [inline drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c]: IS_ERR_OR_NULL check here

File: drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c [hunk 2]
```diff
@@ -2449,28 +2449,36 @@ static int sde_hw_rotator_swts_create(struct sde_hw_rotator *rot)
...
+		}
 	}
 
-	data->srcp_table = dma_buf_map_attachment(data->srcp_attachment,
-			DMA_BIDIRECTIONAL);
-	if (IS_ERR_OR_NULL(data->srcp_table)) {
-		SDEROT_ERR("dma_buf_map_attachment error\n");
-		rc = -ENOMEM;
-		goto err_detach;
+	if (!data->srcp_table) {
```
Comment (kholk) [inline drivers/media/platform/msm/sde/rotator/sde_rotator_r3.c]: IS_ERR_OR_NULL check here too

**Review by kholk — COMMENTED** [review]
Please apply the requested changes, then let's merge this one.

**Comment by MarijnS95** [comment]
@kholk Fair point, but then again do we want to keep this lazy initialization? It doesn't seem strictly necessary anymore, but would prevent us from having unreferenced resources sitting around in the extremely unlikely event that calls to `create` and `destroy` are unbalanced.

**Comment by kholk** [comment]
You have two choices:
1. Create synthetic test cases and make a strict validation of the entire tangled code; or
2. Better lazy-safe than sorry.

You got what I chose. Would you choose the option N.1? :p

**Comment by MarijnS95** [comment]
Sure. The point being that this lazy initialization was ~ necessary ~ in the previous version of these patches, before #2234 came around. Now it's more-or-less cleanup than anything, we have no confirmation that this odd case I just made up ever happens.

Let's go for ⚠️ SAFETY ⚠️ 🎉 🎉 🎉 🎉 🎉

**Pull request merged**